src/network/local_socket_reply.cpp

```diff
@@ -141,36 +139,113 @@ void mp::LocalSocketReply::send_request(const QNetworkRequest& request, QIODevic
     // Build the HTTP method part
     http_data += op;
     http_data += ' ';
-    http_data += request.url().toString();
+    http_data += request.url().path();
+
+    if (request.url().hasQuery())
+    {
+        http_data += "?";
+        http_data += request.url().query();
+    }
+
     http_data += " HTTP/1.1\r\n";
 
     // Build the HTTP Host header
-    // Host can be anything, so we'll use 'multipass'
-    http_data += "Host: multipass\r\n";
+    auto host = request.url().host();
+    if (!host.isEmpty())
+    {
+        http_data += "Host: ";
+        http_data += host;
+        http_data += "\r\n";
+    }
 
     // Build the HTTP User-Agent header
-    http_data += "User-Agent: Multipass/";
-    http_data += mp::version_string;
-    http_data += "\r\n";
+    auto user_agent = request.header(QNetworkRequest::UserAgentHeader).toByteArray();
+    if (!user_agent.isEmpty())
+    {
+        http_data += "User-Agent: ";
+        http_data += user_agent;
+        http_data += "\r\n";
+    }
+
+    local_socket->write(http_data);
 
     if (op == "POST" || op == "PUT")
     {
-        http_data += "Content-Type: application/x-www-form-urlencoded\r\n";
+        http_data = "Content-Type: ";
+        http_data += request.header(QNetworkRequest::ContentTypeHeader).toByteArray();
+        http_data += "\r\n";
 
         if (outgoingData)
         {
+            auto content_length = request.header(QNetworkRequest::ContentLengthHeader).toByteArray();
+            if (!content_length.isEmpty())
+            {
+                http_data += "Content-Length: ";
+                http_data += content_length;
+                http_data += "\r\n";
+            }
+
+            auto transfer_encoding = request.rawHeader("Transfer-Encoding");
+            if (!transfer_encoding.isEmpty())
+            {
+                http_data += "Transfer-Encoding: ";
+                http_data += transfer_encoding;
+                http_data += "\r\n";
+            }
+
+            http_data += "\r\n";
+
+            local_socket->write(http_data);
+            local_socket->flush();
+
             outgoingData->open(QIODevice::ReadOnly);
+            QByteArray data_buffer;
+            const auto max_bytes{32768};
 
-            http_data += "Content-Length: ";
-            http_data += QByteArray::number(outgoingData->size());
-            http_data += "\r\n\r\n";
-            http_data += outgoingData->readAll();
+            while (true)
+            {
+                auto bytes_available = outgoingData->bytesAvailable();
+                if (bytes_available == 0)
+                    break;
+
+                data_buffer = outgoingData->read(bytes_available);
+
+                auto data_ptr = data_buffer.data();
+
+                while (bytes_available > 0)
+                {
+                    auto bytes_to_write = bytes_available < max_bytes ? bytes_available : max_bytes;
```

Comment:
This could be a bit more concise / standard with [`qint64 QIODevice::read(char *data, qint64 maxSize)`](https://doc.qt.io/qt-5.9/qiodevice.html#read), no? You wouldn't need two loops, either?

Also feels weird to `while (true)` and then `break` just a couple lines below.

Reply:
Again, I'm just following how Qt does this in their internal code with some tweaks for our purposes.  Also, in `QHttpMultiPart`, it's possible for a part to be backed by a `QFile`, so the ensuing `QIODevice` here may have much more data that can be read at any one time than can be written to the local socket.  I suppose as you say we can just read and then write `max_bytes` at a time, but that will hurt performance.

Reply:
The file thing is an argument indeed, but then what about memory? Will a file-backed `QIODevice` have the whole file in `::bytes_available()`? We don't want to read it all into memory, do we? So we should still have a max read?

Don't we have a overflow here, btw? If `bytesAvailable()` can be more than the size of `data_buffer`?

Reply:
>  Will a file-backed `QIODevice` have the whole file in `::bytes_available()`?

From my experience, yes.

> We don't want to read it all into memory, do we?

It will be temporary and most of the images are in the 100's of MB's, so I don't think it would impact users in such a way that they would notice, but if cloud images grow to GB's, then that could be a problem.

> So we should still have a max read?

Need to weigh the performance vs. memory usage as mentioned above.  I don't see it as a problem now, but can't guarantee the future.  Perhaps trying to find a `max_bytes` size like 100 MB or something will keep performance and still avoid memory pressure in the future.

> Don't we have overflow here, btw? If `bytesAvailable()` can be more than the size of `data_buffer`?

We are assigning the `QByteArray` `data_buffer` to the `QByteArray` returned by `read()`, so the `read` is doing all of the `QByteArray` allocation.  If we do indeed do this, I should probably ensure the size of `data_buffer` does match `bytes_available` so it is the expected size.

Reply:
> It will be temporary and most of the images are in the 100's of MB's, so I don't think it would impact users in such a way that they would notice, but if cloud images grow to GB's, then that could be a problem.

That can easily be the case for images, so I do believe this to be a problem.



> Need to weigh the performance vs. memory usage as mentioned above. I don't see it as a problem now, but can't guarantee the future. Perhaps trying to find a `max_bytes` size like 100 MB or something will keep performance and still avoid memory pressure in the future.

I really don't think the additional loops will hurt performance here… it's I/O that's going to be the bottleneck here. I'm not saying we should be reading in 16kB chunks, but even 100MB seems excessive. There's just no point in keeping that much in memory for any period of time.

> We are assigning the `QByteArray` `data_buffer` to the `QByteArray` returned by `read()`, so the `read` is doing all of the `QByteArray` allocation. If we do indeed do this, I should probably ensure the size of `data_buffer` does match `bytes_available` so it is the expected size.

Right, sorry, was totally thinking `data_buffer` is of `max_bytes` size… :roll_eyes:

Reply:
Reworked all of this and should be better and more straightforward now.